### ISSUE TITLE
sendtoaddress ignores conf_target (spending whole UTxO with subtractfeefromamount=true)

### ISSUE THREAD

**Issue by whitslack** [bitcoin/bitcoin#16072]
```
$ bitcoind --version | head -n1
Bitcoin Core Daemon version v0.18.0.0-gentoo

$ bitcoin-cli estimatesmartfee 288 ECONOMICAL
{
  "feerate": 0.00001015,
  "blocks": 288
}

$ bitcoin-cli sendtoaddress bc1… ##amount## '' '' true true 288 ECONOMICAL
37############################################################3e

$ bitcoin-cli gettransaction 37…3e | fgrep fee | head -n1
  "fee": -0.00007924,
```

The redacted `##amount##` above was the exact amount of a single unspent output in my wallet (shown by `listunspent`). The resulting transaction (`37…3e`) was 191 bytes in length (virtual size of 110 vbytes or 437 weight units — whatever that means).

So `estimatesmartfee` told me I should have paid a fee of 1.015 sat/byte for confirmation in 288 blocks, but then `sendtoaddress` went ahead and paid 41.487 sat/byte. Not surprisingly, the transaction confirmed in the very next block. I was aiming for it to confirm closer to 288 blocks later, to save on fees.

This is not the first time this has happened to me. My usual routine is that I check the fee needed for a given confirmation target with `estimatesmartfee`, observe that the fee is reasonable, then specify the same confirmation target to `sendtoaddress`. In this case and others, I end up paying an exorbitant fee for an immediate confirmation.

I am not certain whether this behavior always occurs with `subtractfeefromamount=true` or whether it ever occurs with `subtractfeefromamount=false`. I do know that I have almost never had BitcoinD choose a fee rate that has resulted in a confirmation time anywhere close to as long as I requested.

**Comment by VoxR4710** [comment]
I have also observed strange behaviour from this parameter and i am losing my freaking mind by now !
I made a post about it on stack here : https://bitcoin.stackexchange.com/questions/88001/subtractfeefromamount-not-consistently-working

I am starting to think after reading your post that it's just fundamentally broken and that we will have no choice but to do it the long way.

**Comment by laanwj** [comment]
At a cursory glance it looks like the argument is at least used, it's passed [into the coin control object](https://github.com/bitcoin/bitcoin/blob/master/src/wallet/rpcwallet.cpp#L430):
```c++
    if (!request.params[6].isNull()) {
        coin_control.m_confirm_target = ParseConfirmTarget(request.params[6], pwallet->chain().estimateMaxBlocks());
    }
```
It doesn't look like there is any functional test that tests, or even uses this functionality, though, so it might be bugged and simply going untested.

**Comment by instagibbs** [comment]
I'll take a look later today.

**Comment by instagibbs** [comment]
@whitslack @VoxR4710 would either of you have detailed debug.log recordings that you could privately share with me?

I can not replicate in test environment that I set up in `feature_fee_estimation.py` to have non-minimum relay feerate values.

**Comment by whitslack** [comment]
@instagibbs: No, sorry. My `debug.log` begins on 11 June, which is after I opened this issue and switched to using `bitcoin-cli settxfee "$(bitcoin-cli estimatesmartfee …)"` to manage my fee rate rather than trying to use the buggy `conf_target` parameter of `sendtoaddress`.

It should be pretty easy to reproduce on mainnet whenever there's a backlog. Just pick any UTxO from your `listunspent` and do a `sendtoaddress` with the exact amount of the UTxO, passing `true` for `subtractfeefromamount` and some very long `conf_target` like 144 blocks. Compare the fee rate you actually paid against the fee rate reported by `estimatesmartfee 144`.

**Comment by instagibbs** [comment]
Ah I missed the "spend whole coin" condition from the description, let me try again.

**Comment by instagibbs** [comment]
@whitslack notice anything off about the test I wrote? https://github.com/bitcoin/bitcoin/pull/16972 I am unable to reproduce

**Comment by whitslack** [comment]
@instagibbs: Your test case isn't exactly the case I described. I spent only one UTxO, not my entire balance. (I didn't check the rest of the code to see whether your test wallet would have only one UTxO at that point.) Also, I spent a confirmed UTxO, whereas you appear to be spending unconfirmed UTxOs. I don't know if either of those discrepancies are relevant.

**Comment by instagibbs** [comment]
@whitslack the 2nd and 3rd send are spending a single utxo each, which is coincidentally the whole balance. I can change it to mine a block each time as well but I don't expect this to change anything(unconfirmedness doesn't do anything but possibly disqualify it from coin selection in some cases/settings)

**Comment by whitslack** [comment]
@instagibbs: Oh, right; the entire balance is merged into one UTxO by the first spend.

Is there a code-path difference between spending the entire wallet balance and spending the exact amount of only one UTxO among many? I would think that the coin selection algorithm might take a different path to pick one UTxO of exactly the requested amount versus all UTxOs in the wallet.

Let's lay this on the table for now. Next time I have a need to move a whole UTxO to another wallet, I'll try to use `conf_target` instead of `settxfee` and see if I end up overpaying again.

**Comment by Sjors** [comment]
#11413 by @kallewoof adds a whole of bunch of tests for fee handling in `sendtoaddress`. Perhaps those can be adjusted to cover the behavior on master.

**Comment by whitslack** [comment]
I just tested this with my main wallet on v0.18.0, and the behavior was correct. Possibly this issue has already been resolved. I'll switch back to relying on `conf_target` rather than `settxfee` in my regular operations and see if I run into any more occurrences of the misbehavior.

That said, fee estimation has some seemingly bizarre cusps. I've seen cases where `estimatesmartfee 48` returned a feerate of 1 sat/vB but `estimatesmartfee 47` returned a feerate of 20 sat/vB. The function does not seem to be very smooth at all, and it's far from realistic. But that's a whole separate issue.

**Comment by instagibbs** [comment]
It's using bucket estimation, it doesn't try to smooth it out.

**Comment by MarcoFalke** [comment]
> Possibly this issue has already been resolved. 

Ok, closing for now. Please let us know if this happens again.

**Comment by whitslack** [comment]
@MarcoFalke: Please reopen. This is still an issue, and it's still costing me money that I didn't want to spend.

Using Bitcoin Core 0.20.0…

```
$ bitcoin-cli estimatesmartfee 720 ECONOMICAL
{
  "feerate": 0.00001000,
  "blocks": 720
}

$ bitcoin-cli sendmany '' '{"##redacted_A##":#.########,"##redacted_B##":#.########,"##redacted_C##":#.########,"##redacted_D##":#.########}' 0 '' '["##redacted_D##"]' true 720 ECONOMICAL
d…##redacted_txid##…0
```

The resulting transaction had a virtual size of 228 vbytes and paid 2244 satoshis as a fee, which is a rate of **9.9 sat/vB, about *ten times* the fee rate I expected/intended!**

Of note:
* I specified exactly one of the output addresses in the `subtractfeefrom` array.
* The sum of my specified output amounts was exactly equal to the value of the single unlocked UTxO in my wallet.

**Comment by instagibbs** [comment]
@whitslack as a stop-gap you can try using the explicit feerate options now available in master, will be available in 0.21 release:  https://github.com/bitcoin/bitcoin/pull/11413

I'll take another crack at making a test that fails.

**Comment by instagibbs** [comment]
Wrote another test case for sendmany, it's not triggering anything sorry. If you really want to get this fixed I think you're going to have to send me(or someone that does wallet stuff) detailed logs.

**Comment by whitslack** [comment]
> as a stop-gap you can try using the explicit feerate options now available in master, will be available in 0.21 release: #11413

@instagibbs: The new options look promising, though I have a sneaking suspicion that they'll be ignored by the same code path that presently ignores the estimator-based fee options.

> Wrote another test case for sendmany, it's not triggering anything sorry.

Thanks for your efforts. It's strange that this is so trivially reproducible for me using my production mainnet node. This recent occurrence with `sendmany` is not the only time it's happened for me on 0.20.0.

> If you really want to get this fixed I think you're going to have to send me(or someone that does wallet stuff) detailed logs.

Sure. Would I collect detailed enough logs with `-debug=walletdb -debug=estimatefee -debug=selectcoins`? Would I need to add `-debug=coindb` too?

**Comment by instagibbs** [comment]
I *think* most of the wallet/fee related stuff is printed always. Look for `Fee Calculation` string in debug.log? Might be a starting point.

Alternative-alternative for you is to use raw interface and then cross-check that the feerate is right(ugly yes, but wrapping it means you can avoid bugs like this)

**Comment by whitslack** [comment]
> I _think_ most of the wallet/fee related stuff is printed always. Look for `Fee Calculation` string in debug.log? Might be a starting point. My e-mail is in github profile.

I found the `Fee Calculation` for the transaction in question:

```
2020-08-11T01:02:57Z [default wallet] Fee Calculation: Fee:2244 Bytes:228 Needed:2244 Tgt:720 (requested 720) Reason:"Mempool Min Fee" Decay 0.99931: Estimation: (0 - 1000) 100.00% 3390.8/(3390.8 0 mem 0.0 out) Fail: (-1 - -1) -nan% 0.0/(0.0 0 mem 0.0 out)
```

The `Tgt:720` would seem to have come from my `conf_target` parameter, but I don't know what the numbers in `Estimation:` mean. I did show the output from `estimatesmartfee 720 ECONOMICAL`, which showed an estimated fee rate of 1 sat/vB, so I don't know why the fee calculation would have calculated 2244 satoshis for a 228-vB transaction.

> Alternative-alternative for you is to use raw interface and then cross-check that the feerate is right(ugly yes, but wrapping it means you can avoid bugs like this)

I wasn't aware of the `subtractFeeFromOutputs` and `replaceable` arguments to `fundrawtransaction`. Thanks for the pointer. I will most likely use this interface in the future, as it looks easier than locking every UTxO in the wallet and then unlocking the one I want to spend with `sendtoaddress`/`sendmany`.

**Comment by instagibbs** [comment]
> Reason:"Mempool Min Fee"

errr are you sure you aren't using a really small mempool or something?

`getmempoolinfo` <--- results of this?

**Comment by whitslack** [comment]
Current values…

```
$ bitcoin-cli getmempoolinfo
{
  "loaded": true,
  "size": 18232,
  "bytes": 8777962,
  "usage": 35987504,
  "maxmempool": 64000000,
  "mempoolminfee": 0.00003503,
  "minrelaytxfee": 0.00001000
}
```

Regardless of my own mempool's configuration, `sendtoaddress`/`sendmany` ought not to silently inflate the fees they're paying when I have explicitly specified a fee rate. Expected behavior would be to pay the actual fee rate I specified and always relay my own transactions to my peers, even if my transactions can't stick around in my mempool due to memory constraints. An alternative, acceptable (though not desirable) behavior would be to reject my transaction with an error saying it doesn't pay a great enough fee rate to be accepted into my mempool. The current behavior of silently paying many times my explicitly specified fee rate is unexpected and unacceptable and undermines my trust in the software to keep my money safe.

**Comment by instagibbs** [comment]
Maybe, that's an API break. I believe the explicit feerate option actually fails in this case. I'd suggest doing manual transaction creation(walletcreatefundedpsbt) or use explicit feerate.

> "maxmempool": 64000000

64 MB mempool could easily fill during these periods. Mystery solved.

**Comment by whitslack** [comment]
Thanks. Looking forward to the new explicit fee-rate parameters.

**Comment by whitslack** [comment]
For what it's worth, there is no "safe" maximum mempool size. If you say 64 MB is too small, then I say 1 GB is also "too small," and indeed *any* size is too small. Under an eclipse attack, an attacker could easily fill a victim node's mempool of *any* size with valid, high-fee-rate transactions, which pose no risk of confirming and costing the attacker money since the victim won't be able to relay them to any miners due to having been eclipsed from the network.

**Comment by instagibbs** [comment]
The "fix" I'd look at is simply having the block estimator return the max of min mempool entry and it's estimation. That actually seems better than current and won't break anyone's usage?

**Comment by whitslack** [comment]
That could be good. Then front-ends that query the estimator to show a fee rate to the user would have a more accurate idea about what will happen when they actually submit a transaction using `sendtoaddress`/`sendmany`.

My concern would be for Lightning nodes, which need to agree with their peers about the current state of the fee market. If the responses from Bitcoin Core's fee estimator are being influenced by mempool memory constraints, this could trigger mass unilateral channel closures, which is **VERY** expensive.

**Comment by kallewoof** [comment]
I don't believe the explicit fee rate option cares about current mempool limits, it only cares about the hard coded minimum 1 satoshi/byte limit. (I could be mistaken here -- we went back and forth on this during review.)

**Comment by instagibbs** [comment]
Looks like it will simply fail: https://github.com/bitcoin/bitcoin/blob/master/src/wallet/wallet.cpp#L2782

though I don't see a test for it necessarily

**Comment by rebroad** [comment]
> > Reason:"Mempool Min Fee"
> 
> errr are you sure you aren't using a really small mempool or something? The wallet will definitely give you something higher if what you give doesn't hit your mempool minimum.
> 
> `getmempoolinfo` <--- results of this?

Surely the mempool minimum shouldn't come into the equation at all...

**Comment by whitslack** [comment]
> Surely the mempool minimum shouldn't come into the equation at all...

@rebroad: Sadly, it does, and the reason is one of privacy. If your node isn't relaying any transactions paying fees less than *X* sat/vB because its mempool is full, but then it relays one transaction that pays a fee of *X*/2 sat/vB, then it would be obvious to your peers that your node originated that transaction.

Personally, I wish there were an option to say, "I care more about cost than I care about privacy; transmit my cheap transaction anyway," but such an option does not exist.
